bool done
float32 time_taken

